### ISSUE TITLE
iSTFT produces RuntimeError with center=False and Blackman/Bartlett/Hann windows

### ISSUE THREAD

**Issue by fakufaku** [pytorch/pytorch#91309]
### 🐛 Describe the bug

When using the `torch.istft` function with the option `center=False` and windows produced by `torch.blackman_window`, `torch.hann_window` or `torch.bartlett_window`, then a `RuntimeError` is produced with message
`window overlap add min: 1`.

The expected behavior is that no error is produced.

```python
import math
import torch


f = 1500.0  # exactly periodic
fs = 48000
nfft = 1024
hop = 256
dtype = torch.float64

original = torch.sin(2 * math.pi * f / fs * torch.arange(2 * fs, dtype=dtype))

# torch
# - use `center=False`
# - zero-pad the front of the signal
windows = {
    "hamming": torch.hamming_window(nfft, dtype=dtype),
    "kaiser": torch.kaiser_window(nfft, dtype=dtype),
    "blackman": torch.blackman_window(nfft, dtype=dtype),
    "bartlett": torch.bartlett_window(nfft, dtype=dtype),
    "hann": torch.hann_window(nfft, dtype=dtype),
}

for center in [True, False]:
    for win_name, win in windows.items():
        # center == True
        try:
            PT = torch.stft(
                original,
                n_fft=nfft,
                hop_length=hop,
                window=win,
                return_complex=True,
                pad_mode="constant",
                center=center,
            )
            recon_pt = torch.istft(
                PT, n_fft=nfft, hop_length=hop, window=win, center=center
            )
            err = float(abs(recon_pt - original[: recon_pt.shape[0]]).max())
            print(
                f"{win_name=:9s} {center=} {err=:.6e}",
            )
        except RuntimeError as e:
            print(
                f"{win_name=:9s} {center=} Exception!!!",
            )
            print(e)
```

Output
```
win_name=hamming   center=True err=7.771561e-16
win_name=kaiser    center=True err=7.771561e-16
win_name=blackman  center=True err=6.661338e-16
win_name=bartlett  center=True err=5.551115e-16
win_name=hann      center=True err=6.661338e-16
win_name=hamming   center=False err=6.217249e-15
win_name=kaiser    center=False err=3.472223e-12
win_name=blackman  center=False Exception!!!
istft(CPUComplexDoubleType[513, 372], n_fft=1024, hop_length=256, win_length=1024, window=torch.DoubleTensor{[1024]}, center=0, normalized=0, onesided=None, length=None, return_complex=0) window overlap add min: 1
[ CPUBoolType{} ]
win_name=bartlett  center=False Exception!!!
istft(CPUComplexDoubleType[513, 372], n_fft=1024, hop_length=256, win_length=1024, window=torch.DoubleTensor{[1024]}, center=0, normalized=0, onesided=None, length=None, return_complex=0) window overlap add min: 1
[ CPUBoolType{} ]
win_name=hann      center=False Exception!!!
istft(CPUComplexDoubleType[513, 372], n_fft=1024, hop_length=256, win_length=1024, window=torch.DoubleTensor{[1024]}, center=0, normalized=0, onesided=None, length=None, return_complex=0) window overlap add min: 1
[ CPUBoolType{} ]
```

### Versions

PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6.1 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.8 (main, Nov 24 2022, 08:08:27) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-12.6.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.4
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[conda] numpy                     1.23.4          py310hb93e574_0
[conda] numpy-base                1.23.4          py310haf87e8b_0
[conda] torch                     1.13.1                   pypi_0    pypi
[conda] torchaudio                0.13.1                   pypi_0    pypi

**Comment by drisspg** [comment]
Error is stemming from this line of code: https://github.com/pytorch/pytorch/blob/4437d0d161f71620c6f5a601ba3955e0b42a17ce/aten/src/ATen/native/SpectralOps.cpp#L1150

**Comment by drisspg** [comment]
https://pytorch.org/docs/stable/generated/torch.istft.html?highlight=istft

**Comment by fakufaku** [comment]
I think I understand where the problem is coming from. It is indeed related to the note you link to, that is the envelope becomes zero. In this case, Hann, Blackman, and Bartlett windows all have a zero in first position. 
* When `center=False`, this means that the first coefficient of the signal cannot be recovered.
* When `center=True`, the same thing happen, but since the first sample was zero padded anyway, this phenomenon is ignored.

I believe however that throwing an error is not the appropriate behavior for two reasons.

1. As the solution of a least-square problem, the iSTFT is well defined even when the envelope becomes zero. The time samples where the envelop become zero need to satisfy the equation `0 x = 0`, that is, they could take any value. Setting `x=0` will provide the minimum norm solution.
2. If the envelope contains some zeros and the same window was used in the forward STFT, then those samples have been zeroed out by the forward window anyway.

I would suggest the following fix. When the envelope is zero, rather than throwing an error, the coefficients corresponding to the zero envelope are set to zero.

**Comment by drisspg** [comment]
Do you know if this proposed behavior matches that of librosa? Would you be interested in creating a PR with these changes?

**Comment by fakufaku** [comment]
It looks like it does. This is the relevant section of librosa's [istft code](https://github.com/librosa/librosa/blob/main/librosa/core/spectrum.py#L597)

```python
    # Normalize by sum of squared window
    ifft_window_sum = window_sumsquare(
        window=window,
        n_frames=n_frames,
        win_length=win_length,
        n_fft=n_fft,
        hop_length=hop_length,
        dtype=dtype,
    )

    if center:
        start = n_fft//2
    else:
        start = 0

    ifft_window_sum = util.fix_length(ifft_window_sum[..., start:], size=y.shape[-1])

    approx_nonzero_indices = ifft_window_sum > util.tiny(ifft_window_sum)

    y[..., approx_nonzero_indices] /= ifft_window_sum[approx_nonzero_indices]

    return y
```

Actually, librosa's solution is to just not divide those coefficients where the window sum is zero 😄 This is equivalent to what I propose since those coefficients would already be set to zero by the application of the window itself.

Sure, I could look into creating a PR!

**Comment by fakufaku** [comment]
The fix may be to remove lines 1147 to 1152, and replace [line 1154 of SpectralOps.ccp](https://github.com/pytorch/pytorch/blob/4437d0d161f71620c6f5a601ba3955e0b42a17ce/aten/src/ATen/native/SpectralOps.cpp#L1154)
```c++
y = (y / window_envelop);  // size: (channel, expected_output_signal_len)
```
by something like
```c++
y = at::where(window_envelop.abs().lt(1e-11), 0, y / window_envelop);  // size: (channel, expected_output_signal_len)
```
Do you think that would that work ?

**Comment by fakufaku** [comment]
Hum, it looks like this has come up before and was rejected on the basis that perfect reconstruction is then not possible.
pytorch/pytorch#62323
In my opinion this prevents some use case when you want to handle the padding yourself. @nateanl what do you think ?

**Comment by nateanl** [comment]
I think to ignore the error, we can add an extra argument like "allow_lossy" and set it to `False` by default, when enabling it, the function won't throw error when window envelop has zero values.

In this use case, is perfect reconstruction not a hard requirement? If not, I'm wondering what's the unique feature of custom padding compared to setting `center=True`, since you have to do padding by yourself in order to keep the perfect reconstruction?

**Comment by fakufaku** [comment]
I guess a new argument is a possible solution.

> In this use case, is perfect reconstruction not a hard requirement?

I suppose things depend on the definition of `stft` `istft`. The situation now is that call `torch.stft` with a, e.g., Blackman window will not throw an error even though the signal produced can't be inverted (with `center=False` at least). If the goal is to prevent users from doing non-invertible transformations, this is not completely logical.

Now that I think a little bit deeper about it, there are two ways to test perfect reconstruction of the STFT:
1. `istft(stft(x)) == x`
2. `stft(istft(y)) == y`
I am guessing the first one is used and has led to the current situation. On the other hand, if you use 2, then the tests would pass regardless of the window used.

So in the end it depends what you want to prevent users from doing.

As for the use case for padding, as I understands, `center` will put the time index at the middle of the frame. One use case is if you want to have the time index at the beginning of the frame.

An alternative solution instead of having a hard error and a parameter is to produce a warning when the window is lossy.

**Comment by fakufaku** [comment]
Another test to check the correctness of the `istft` is that the residual should be orthogonal to the reconstructed signal:
```python
import torch

n_fft = 512
hop = 128
n_frames = 126
window = torch.hamming_window(n_fft)
center = True
dtype = torch.complex64

y = torch.zeros((n_fft // 2 + 1, n_frames), dtype=dtype).normal_()
x = torch.istft(y, n_fft, hop_length=hop, window=window, center=False)
y_hat = torch.stft(
    x, n_fft, hop_length=hop, window=window, return_complex=True, center=False
)

error = y - y_hat  # the reconstruction error

# normalize the vectors to check the inner product
error /= torch.linalg.norm(error)
y_hat_norm = y_hat / torch.linalg.norm(y_hat)

# check orthogonality
orth_test = abs((error * y_hat_norm).mean())
print(orth_test)
assert orth_test < 1e-6
```
This should be true regardless of the window, padding, parameters, etc. It is true because the output of the STFT is the solution of the least square problem `istft(y) = \arg \min_x \| y - stft(x) \|^2`.

**Comment by fakufaku** [comment]
I think one more use case for custom padding is if you want to do block-wise stft processing. In that case we don't want to zero pad, but pad with values from the previous block.

**Comment by nateanl** [comment]
I see. zero-padding can potentially affect the performance in block-wise stft processing. For example, in streaming ASR use case, the zero padding in each chunk will introduce gap between chunks which may not be optimal for decoding (cc @hwangjeff) 

Regarding the test if the stft can be invertible, it seems the current check is more strict (which is good?). Is there a solution that can be run in `torch.stft`? We can't run `istft(stft(istft(x)))` inside `torch.stft` function, right?

**Comment by peterbell10** [comment]
@fakufaku In your use case, are you centering the window but need to manually specify the padding values? If so, you could pass `center=True` to the `istft` and it will discard the padding values before checking the NOLA condition.

**Comment by fakufaku** [comment]
> Regarding the test if the stft can be invertible, it seems the current check is more strict (which is good?). Is there a solution that can be run in torch.stft? We can't run istft(stft(istft(x))) inside torch.stft function, right?
@nateanl Ah, I meant this more as a unit test to check the correctness, not something to do at run time. The current check should be fine for that. Only the behavior is a problem (hard fail).

@peterbell10 One important use case is to allow to match the output of other STFT implementations (other than librosa). The option `center=True` does two unrelated things 1. padding, 2. centering the frame time. Other implementation take different choices. With the hard fail, it is currently difficult to match the result given by other implementation (e.g. [pyroomacoustics](https://github.com/LCAV/pyroomacoustics) or [paderbox](https://github.com/fgnt/paderbox)). See [this tweet](https://twitter.com/rdesh26/status/1604769309565235200) for an actual issue.

**Comment by fakufaku** [comment]
@peterbell10 Yes, I understand this is a single operation. What I meant is that it covers both the padding that ensures perfect reconstruction independent of the window function *and* how much to pad with so that the frame is centered. There are other ways to pad so as to get perfect reconstruction.

> My question is do you pad by the same amount (n_fft // 2), or by a different amount?

Many implementations pad by the minimum amount, i.e., `n_fft - hop_length`.

**Comment by hubertsiuzdak** [comment]
If someone is searching for a workaround to handle custom padding in torch.istft, you might want to check out [this implementation](https://github.com/charactr-platform/vocos/blob/03c4fcbb321e4b04dd9b5091486eedabf1dc9de0/vocos/spectral_ops.py#L7-L75).

This workaround essentially reimplements ISTFT using PyTorch operators, but it properly trims the window envelope. It should be straightforward to adapt to other padding schemes. I personally required the typical CNN padding style of `(kernel_size - stride) // 2`. I refer to this as "same", but now I think it can be a bit misleading, as it doesn't maintain exactly the same shape due to strides, but rather upsamples by an integer factor. This is in contrast to `center=True`, which pads/trims too much.

Still feels a bit awkward that this isn't natively supported.

Related issues: https://github.com/pytorch/pytorch/issues/62323, https://github.com/pytorch/pytorch/issues/94718, https://github.com/pytorch/audio/issues/500, https://github.com/pytorch/audio/issues/427

**Comment by fakufaku** [comment]
Any chance we can remove NOLA, add an option to disable it, or provide custom padding length ?

I am running into this issue again. Exactly same scenario as [this issue](https://github.com/pytorch/audio/issues/427).

**Comment by happyTonakai** [comment]
`scipy.signal.check_NOLA` gives `True` for the hann window with 50% overlap. I didn't check the source code of torch's implementation but I think there may be something wrong with it... It's strange that the stft of a signal cannot be inverted by istft with the same parameters.

**Comment by brynluisi** [comment]
This type of padding is very common in many ML projects: 

`        x = F.pad(x, (int((n_fft - self.hop_length) / 2), int((n_fft - self.hop_length) / 2)), mode="reflect")
`

and I believe this implementation should run without error, but doesn't: 

`class Spectrogram(torch.nn.Module):
    def __init__(self, n_fft, hop_length, win_length):
        super(Spectrogram, self).__init__()
        self.n_fft = n_fft
        self.hop_length = hop_length
        self.win_length = win_length
        window = torch.hann_window(self.win_length)
        self.register_buffer("window", window)

    def forward(self, x):
        x = F.pad(x, (int((self.n_fft - self.hop_length) / 2), int((self.n_fft - self.hop_length) / 2)), mode="reflect")

        x = x.squeeze(1)

        X = torch.stft(
            x,
            n_fft=self.n_fft,
            hop_length=self.hop_length,
            win_length=self.win_length,
            window=self.window,
            center=False,
            pad_mode="reflect",
            normalized=False,
            onesided=True,
            return_complex=True,
        )

        return X

    def inverse(self, X):

        x = torch.istft(
            X,
            n_fft=self.n_fft,
            hop_length=self.hop_length,
            win_length=self.win_length,
            window=self.window,
            center=False,
            onesided=True,
            return_complex=False,
        )
        return x`
        
        Has there been any updates on this or known workarounds?
